cantor: fix linking libR.so and add patch which fixes issue with nix-installed julia

hqurve opened the pull request:
###### Description of changes

The libR.so is stored in ${R}/lib/R/lib instead of the usual ${R}/lib. I suspect that this causes issues with the build system and prevents it from linking libR.so correctly.

Fixes #234421

This PR also applies the patch https://invent.kde.org/education/cantor/-/commit/5c038e422c8fe2fcc14ddca2f8ec8d8c6fbca592. This patch removes the backend's reliance on the julia executable on the path. Prior to this patch the julia backend would not allow julia which was installed using for example `nix profile install` since it was installed via a symlink

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

hqurve committed:
cantor: fix linking libR.so

The libR.so is stored in ${R}/lib/R/lib instead of the usual ${R}/lib.
I suspect that this causes issues with the build system and prevents it from
linking libR.so correctly.

Fixes https://github.com/NixOS/nixpkgs/issues/234421

hervyqa commented:
outside of this PR from R, the path for julia is also not detected. even though julia was installed.

> Some requirements are not fulfilled: The path to Julia specified in the application settings must point directly to the executable. Symlinks are not allowed. Please provide the correct path in the application settings and try again.

do i have to create a new issue?

![Screenshot_20230529_035053](https://github.com/NixOS/nixpkgs/assets/45872139/b5dc17c5-8bb1-41ba-b1d1-10a6ea7f6377)

hqurve commented:
> 



> outside of this PR from R, the path for julia is also not detected. even though julia was installed.
> 
> > Some requirements are not fulfilled: The path to Julia specified in the application settings must point directly to the executable. Symlinks are not allowed. Please provide the correct path in the application settings and try again.
> 
> do i have to create a new issue?

No, I'll fix it one time

hqurve committed:
cantor: add patch which removes julia backend's dependence on julia executable in PATH

The cantor julia backend currently does not allow the executable to be a symlink.
This is an issue if julia is installed using nix (which uses symlinks to install files).

The added patch fixes this issue by not depending on the julia executable to be in PATH

hqurve commented:
I tested different methods of installing julia
- using `nix-shell -p julia` to load a shell with julia on the path works
- using `nix profile install nixpkgs#julia` to install julia in my profile does not work (it complains about symlinks as show in the screenshot).

The only mention of `symlink` in the cantor code (excluding localization files) is in https://invent.kde.org/education/cantor/-/blob/master/src/backends/julia/juliabackend.cpp#L84. I believe that this logic was first introduced in https://invent.kde.org/education/cantor/-/commit/1e144559c034ff70b23b3d6b5c1b7b28dbf8af18. 

Since there wasnt an associated explanation and this logic is only present in the julia backend, I tested a build with this logic removed. I believe that it works even if the path is a symlink. I'll try to ask why this logic exists and if it can be safely patched out.

For now, the updated derivation includes the patch

ttuegel reviewed (commented):
Do you know why upstream disallowed links to Julia?

hqurve commented:
According to https://bugs.kde.org/show_bug.cgi?id=470381#c1, this restriction was introduced to ensure that julia's shared libraries were found at runtime.

Your comment prompted me to look at the bug report again and I saw that it was fixed in https://invent.kde.org/education/cantor/-/commit/5c038e422c8fe2fcc14ddca2f8ec8d8c6fbca592. This patch removes the backend's reliance on the julia executable on the path. I've added the patch from upstream.

I've rebuilt cantor and tested a simple 1+1 in both Julia and R. While testing julia, I used the following configurations

- julia not on path
- julia loading with `nix-shell -p`
- julia installed with `nix profile install`

(It shouldnt really matter since julia is no longer used from the path. But I just wanted to be sure)
